libs/chain-signatures/contract/src/lib.rs

```diff
@@ -72,80 +81,82 @@ impl Default for VersionedMpcContract {
 }
 
 #[near(serializers=[borsh])]
-#[derive(Debug, Clone)]
-pub struct AllowedCodeHash {
-    code_hash: CodeHash,
-    added: BlockHeight,
+#[derive(Debug)]
+pub struct TeeState {
+    allowed_tee_proposals: AllowedTeeProposals,
+    historical_tee_proposals: Vec<TeeProposal>,
+    votes: CodeHashesVotes,
+    tee_participant_info: IterableMap<AccountId, TeeParticipantInfo>,
 }
 
-#[near(serializers=[borsh])]
-#[derive(Debug, Default)]
-pub struct AllowedCodeHashes {
-    allowed_code_hashes: Vec<AllowedCodeHash>, // ordered by `start`
+impl Default for TeeState {
+    fn default() -> Self {
+        Self {
+            allowed_tee_proposals: Default::default(),
+            historical_tee_proposals: Default::default(),
+            votes: Default::default(),
+            tee_participant_info: IterableMap::new(StorageKey::TeeParticipantInfo),
+        }
+    }
 }
 
-impl AllowedCodeHashes {
-    /// Removes all expired code hashes and returns the number of removed entries.
-    fn clean(&mut self, current_block_height: BlockHeight) -> usize {
-        // Find the first non-expired entry
-        let expired_count = self
-            .allowed_code_hashes
-            .iter()
-            .position(|entry| entry.added + TEE_UPGRADE_PERIOD >= current_block_height)
-            .unwrap_or(self.allowed_code_hashes.len());
-
-        // Remove all expired entries
-        self.allowed_code_hashes.drain(0..expired_count);
-
-        // Return the number of removed entries
-        expired_count
+impl TeeState {
+    pub fn tee_status(&self, participants: Vec<AccountId>) -> BTreeMap<AccountId, TeeQuoteStatus> {
+        let now_sec = env::block_timestamp_ms() / 1_000;
+        participants
+            .into_iter()
+            .map(|account_id| {
+                let status = self
+                    .tee_participant_info
+                    .get(&account_id)
+                    .map(|tee_participant_info| {
+                        TeeQuoteStatus::from(tee_participant_info.verify_quote(now_sec))
```

Comment:
Don't we also need to ensure that the participant is running a docker image that is is whitelisted?

Reply:
just to bump this comment for visibility. I guess here, we are essentially checking if one of the certificates expired?

Shouldn't we additionally make sure that the docker image hash the participant is running is still valid and in the list of allowed hashes?

Reply:
@pbeza here's the thread.